### PR TITLE
[IMP] crm: replace meeting count by date for lead stat btn

### PR DESCRIPTION
**Pull request by Odoonan**
Instead of having a count field for meetings linked to a crm.lead, show a date instead:
- If there is a next meeting, its date
- Otherwise, show last meeting date
- If no meeting, show no date and display "No Meetings"

This commit removes the field "calendar_event_count" that has no use anymore. A test using the count field is adapted.

UPG PR https://github.com/odoo/upgrade/pull/4935

Task-3418397

### CODE REVIEW

File: addons/crm/models/crm_lead.py [hunk 1]
```diff
@@ -537,13 +537,18 @@ def _compute_recurring_revenue_prorated(self):
...
+        now = fields.Datetime.now()
+        leads_with_meetings = self.filtered('calendar_event_ids')
+        (self - leads_with_meetings).meeting_display_date = False
+        for lead in leads_with_meetings:
+            start_dts = lead.calendar_event_ids.mapped('start')
+            future_start_dts = [dt for dt in start_dts if dt > now]
+            if future_start_dts:
+                lead.meeting_display_date = min(future_start_dts)
+            else:
+                lead.meeting_display_date = max(start_dts)
```
Comment (jeh-odoo) [inline addons/crm/models/crm_lead.py:551]: We could still use a read_group here:
```suggestion
        now = fields.Datetime.now()
        meeting_data = self.env['calendar.event'].sudo()._read_group([
            ('opportunity_id', 'in', self.ids),
        ], ['opportunity_id'], ['start:array_agg', 'start:max'])
        mapped_data = {
            lead: {
                'last_meeting_date': last_meeting_date,
                'next_meeting_date': min([dt for dt in meeting_start_dates if dt > now] or [False]),
            } for lead, meeting_start_dates, last_meeting_date in meeting_data
        }
        for lead in self:
            lead_meeting_info = mapped_data.get(lead)
            if not lead_meeting_info:
                lead.meeting_display_date = False
            elif lead_meeting_info['next_meeting_date']:
                lead.meeting_display_date = lead_meeting_info['next_meeting_date']
            else:
                lead.meeting_display_date = lead_meeting_info['last_meeting_date']
```

**Review by jeh-odoo — COMMENTED** [review]
Hello @Odoonan 
Just one comment and a suggestion for this one 🙂 
Also don't forget to adapt the upgrade script to 16.5
Cheers!

File: addons/crm/views/crm_lead_views.xml [hunk 2]
```diff
@@ -31,9 +31,9 @@
                                 context="{'partner_id': partner_id}"
                                 attrs="{'invisible': [('type', '=', 'lead')]}">
                                 <div class="o_stat_info">
-                                    <field name="calendar_event_count" class="o_stat_value"/>
-                                    <span class="o_stat_text" attrs="{'invisible': [('calendar_event_count', '&lt;', 2)]}"> Meetings</span>
-                                    <span class="o_stat_text" attrs="{'invisible': [('calendar_event_count', '&gt;', 1)]}"> Meeting</span>
+                                    <field name="meeting_display_date" class="o_stat_value" attrs="{'invisible': [('meeting_display_date', '=', False)]}"/>
+                                    <span class="o_stat_text" attrs="{'invisible': [('meeting_display_date', '=', False)]}"> Meetings</span>
+                                    <span class="o_stat_text" attrs="{'invisible': [('meeting_display_date', '!=', False)]}"> No Meetings</span>
```
[the remark this answers is not in the document]
Comment (Odoonan) [reply addons/crm/views/crm_lead_views.xml:36]: Hello @jeh-odoo  -> context_today is a date (~ env / user tz) and I think we want to see next date if we already had a meeting in the morning. We need precision up to datetime, and to compare utc values. Then datetime.datetime.now() could work but I never see it used in attrs. Only in domains or eval on records. I've tried a few things using it but well... It failed ^^' I see in the spec that it works as is. But I can always add another compute. What do you prefer? Cheers :)

[the remark this answers is not in the document]
Comment (jeh-odoo) [reply addons/crm/views/crm_lead_views.xml:36]: I'm a little sad but thanks for having a look. Then yes we could indeed improve a bit the wording here to have something consistent with the other modules. We can go for two computed fields. A `meeting_display_datetime` and a `meeting_display_label` perhaps ? And they can use the same compute method so it shouldn't add much for the diff.
I don't think we need to have a special case for only one meeting like in hr so the No Meeting, Next Meeting and Last Meeting should be sufficient for the purpose here imo.

File: addons/crm/tests/test_crm_lead.py [hunk 3]
```diff
@@ -560,7 +560,7 @@ def test_crm_lead_unlink_calendar_event(self):
                 'stop': '2022-07-13 10:00:00',
             }
         ])
-        self.assertEqual(lead.calendar_event_count, 1)
+        self.assertEqual(len(lead.calendar_event_ids), 1)
```
Comment (tde-banana-odoo) [inline addons/crm/tests/test_crm_lead.py:563]: Could probably add some checks on the new fields while being there, that way it is covered.

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing !

**Comment by Odoonan** [comment]
@tde-banana-odoo Hello,

I've added a small separate test to test the new fields to keep the purpose of existing test better defined.

Cheers! :)

**Comment by tde-banana-odoo** [comment]
@robodoo r+

Thanks for update :)

**Pull request closed without merging**